s2fft/utils/healpix_ffts.py

```diff
@@ -5,10 +5,24 @@
 import torch
 from s2fft.sampling import s2_samples as samples
 from functools import partial
+from s2fft.utils.jax_pritimive import BasePrimitive, register_primitive
+import jaxlib.mlir.ir as ir
+from s2fft_lib import _s2fft
+from jaxlib.hlo_helpers import custom_call
+from jax._src.lib.mlir.dialects import hlo
+from jax._src.interpreters import mlir
+from jax.core import Primitive, ShapedArray
+from jax.interpreters import ad, xla
+from jax.sharding import Mesh, NamedSharding
+from jax.sharding import PartitionSpec as P
+from typing import Tuple
+from jax._src.api import ShapeDtypeStruct
```

Comment:
We should avoid importing anything from `jax._src` as its not part of public API.

Reply:
Not possible for this exact case.
I would look for an alternative
But when writing primitives, we have to dig deep in _src

Reply:
> Not possible for this exact case.
> I would look for an alternative

`ShapeDtypeStruct` is [available under the top-level `jax` namespace](https://jax.readthedocs.io/en/latest/_autosummary/jax.ShapeDtypeStruct.html).

> But when writing primitives, we have to dig deep in _src

I don't think this is true as for example [the `kepler_jax.py` file](https://github.com/dfm/extending-jax/blob/main/src/kepler_jax/kepler_jax.py) in the `extending-jax` tutorial repository does not have any imports from under the `jax._src` subpackage.

Reply:
I am gonna make a second pass on the file and clean up
But I started from my own style of primitives see [here](https://github.com/DifferentiableUniverseInitiative/jaxDecomp/blob/main/jaxdecomp/_src/transpose.py)
Which itself is inspired from nvidia style [here](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/cpp_extensions/normalization.py#L40)
Which is more up to date and also has the nessecary language to define multi-gpu primitives that DFM's example does not do.
Some of the imports are used as argument type hints .. for [example in my code](https://github.com/DifferentiableUniverseInitiative/jaxDecomp/blob/main/jaxdecomp/_src/transpose.py#L153)
Whereas in NVIDIA's [code](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/cpp_extensions/normalization.py#L185)  it is not used (so no need for import)

Nevertheless, I will make a pass and remove `_src` imports